threadpoolctl.py

```diff
@@ -599,6 +211,520 @@ def get_original_num_threads(self):
+                if not ps_api.GetModuleFileNameExW(
+                        h_process, h_module, ctypes.byref(buf),
+                        ctypes.byref(n_size)):
+                    raise OSError("GetModuleFileNameEx failed")
+                filepath = buf.value
+
+                # Store the module if it is supported and selected
+                self._make_module_from_path(filepath)
+        finally:
+            kernel_32.CloseHandle(h_process)
+
+    def _make_module_from_path(self, filepath):
+        """Store a module if it is supported and selected"""
+        # Required to resolve symlinks
+        filepath = self._realpath(filepath)
+        # `lower` required to take account of OpenMP dll case on Windows
+        # (vcomp, VCOMP, Vcomp, ...)
+        filename = os.path.basename(filepath).lower()
+
+        # Loop through supported modules to find if this filename corresponds
+        # to a supported module.
+        for _, candidate_module in _SUPPORTED_MODULES.items():
+            # check if filename matches a supported prefix
+            prefix = self._check_prefix(filename,
+                                        candidate_module["filename_prefixes"])
+
+            # filename does not match any of the prefixes of the candidate
+            # module. move to next module.
+            if prefix is None:
+                continue
+
+            # filename matches a prefix. Check if it matches the request. If
+            # so, create and store the module.
+            if (prefix in self.prefixes or
+                    candidate_module["user_api"] in self.user_api):
+                module_class = globals()[candidate_module["module_class"]]
+                module = module_class(filepath, prefix)
+                self.modules.append(module)
+
+    def _check_prefix(self, library_basename, filename_prefixes):
+        """Return the prefix library_basename starts with
+
+        Return None if none matches.
+        """
+        for prefix in filename_prefixes:
+            if library_basename.startswith(prefix):
+                return prefix
+        return None
+
+    @classmethod
+    def _get_libc(cls):
+        """Load the lib-C for unix systems."""
+        libc = cls._system_libraries.get("libc")
+        if libc is None:
+            libc_name = find_library("c")
+            if libc_name is None:  # pragma: no cover
+                return None
+            libc = ctypes.CDLL(libc_name, mode=_RTLD_NOLOAD)
+            cls._system_libraries["libc"] = libc
+        return libc
+
+    @classmethod
+    def _get_windll(cls, dll_name):
+        """Load a windows DLL"""
+        dll = cls._system_libraries.get(dll_name)
+        if dll is None:
+            dll = ctypes.WinDLL("{}.dll".format(dll_name))
+            cls._system_libraries[dll_name] = dll
+        return dll
+
+    @classmethod
+    def _realpath(cls, filepath, cache_limit=10000):
+        """Small caching wrapper around os.path.realpath to limit system calls
```

Comment:
Maybe make it explicit that we use a cls level cache instead of an instance level cache because we can safely assume that the filepath of loaded charged library will never change during the lifetime of a program.